enhancements/microshift/microshift-pluggable-core-components.md

```diff
@@ -0,0 +1,356 @@
+
+The following changes would be introduced to allow the described type of modularity:
+
+Refactor and reuse the kustomize loader into a new module for core components which
+would load components like:
+
+`/usr/lib/microshift/components.d/004-multus/kustomize.yaml`
+`/usr/lib/microshift/components.d/005-networking-ovnk/kustomize.yaml`
+`/usr/lib/microshift/components.d/010-storage-topolvm/kustomize.yaml`
+
+  MicroShift component manager would scan the `/usr/lib/microshift/components.d/`
+  directory executing the component manifests in alphabetical order.
+
+  Today for the CNI/CSI components there is no wait or readiness conditions,
+  if this needed to be implemented in the future it could be defined in yaml
+  metadata.
+
+  `/etc/microshift/components.d/` would also be provided to allow passing down
+  specific configuration down to components; for example:
+
+  `/etc/microshift/components.d/010-storage-topolvm/config.yaml`
+
+   such yaml would parsed and passed down on the templates of the following
+   point as `{{ .config.value }}` or `{{ .config.sub.value }}`.
+
+*  Kustomize yamls would be pre-processed with go templates to allow insertion of
+  variables as we do today for core assets, i.e.: `{{ .KubeconfigDir }}`,
+  `{{ .BaseDomain }}`,
+  [etc...](https://github.com/openshift/microshift/blob/eb0488fd167333256c69f4eeedc5463c4035a532/pkg/components/render.go#L24).
+  Image references will be injected in the `images` section
+  of the kustomize format not being embedded in `microshift/pkg/releases.go`
+  or passed down as template variables anymore.
+
+* The functionality used in the topo-lvm component to embed the lvmd config file
+  in a config map could be performed with a config map generator, which would contain
+  a default config file. The default config file could be overriden via a custom component
+  variable.
+
+
+The following changes would be introduced to packaging:
+
+* Convert `microshift` into a virtual package that depends on `microshift-core`,
+  `microshift-networking-ovnk`, `microshift-storage-topolvm`.
+
+* The `microshift-networking-ovnk` package would include the necessary
+  `/usr/lib/microshift/components.d/05-networking-ovnk/` yamls, including crds.
+
+* The `microshift-storage-topolvm` package would include the necessary
+  `/usr/lib/microshift/components.d/10-storage-topolvm/` yamls, including crds.
+
+* The rebase script should be updated to extract those core components
+  as kustomize yamls instead of embedded assets.
+
+* Incompatible component rpms should be impossible to install together.
+  i.e. `microshift-networking-bridge` plus `microshift-networking-ovnk`,
+  this can be handled with Conflict spec clauses. (This is just an example,
+  since microshift-networking-bridge would not be necessary as it's the
+  default for cri-o)
+
+### Workflow Description
+
+#### Deploying
+
+The workflow for deployment is already described in the `User Stories` section.
```

Comment:
Not anymore. You'll probably want to refer to the beginning of the proposal section instead.

Reply:
oops :D